prov/verbs: Track ep state to prevent duplicate shutdown events

shefty opened the pull request:
Track the connection state of msg endpoints.  This is needed to safely transition into a disconnect state, while ensuring that we only report a shutdown event to the user once.

There are 2 separate ways that we can determine that an endpoint has been disconnected.  One is by reading a disconnect event from the rdma cm.  Another is by reading an error completion from the CQ associated with the underlying QP.  Because we're not guaranteed to get a disconnect event (the remote peer system may have crashed without generating an RDMA disconnect traffic), if we read an error completion from the CQ, we need to notify the user that the local EP is no longer usable.  We do this by generating a local shutdown event.

However, we _could_ read a disconnect event from the EQ.  We have no way of knowing beforehand, or which order the disconnect notification could occur (from a CQ or EQ event).  The state checks allow us to transition the EP once and generate a single local shutdown event.

shefty commented:
This targets a fix for #9123, but the changes are untested at the moment.  (Verbs wasn't enabled on the cluster I was developing on.)  My larger concern is that there is no locking through many of these calls.  That may not be an issue for some of the CM paths.  The app can call multiple CM calls on the same endpoint in parallel.  But the EQ handling does need some sort of coordination with reading the CQ.

The end goal is to use similar locking as implemented by the tcp provider (progress level locks at the domain level), but that's not yet in place.  Posting for early review/feedback.

chien-intel commented on prov/verbs/src/verbs_cq.c:
```diff
@@ -227,7 +227,7 @@ int vrb_poll_cq(struct vrb_cq *cq, struct ibv_wc *wc)
 		wc->wr_id = (uintptr_t) ctx->user_ctx;
 		if (wc->status != IBV_WC_SUCCESS && wc->status != IBV_WC_WR_FLUSH_ERR)
-			vrb_shutdown_qp_in_err(ctx->ep);
+			vrb_shutdown_ep(ctx->ep);
```
I don't know what non-IBV_WC_SUCCESS would result in QP transitioning to IBV_QPS_ERR by the adapter.  That's the reason behind ibv_query_qp to make sure we are only closing on QP in IBV_QPS_ERR.  This would potentially close QP that's still in RTS with a non-fatal status.  Also, don't think status and behavior would be the same between HW vendors thus checking current qp state is the only safe way to do this.

shefty commented:
Update on locking: Until we can integrate rdm support directly into the provider and rework locking, we're in a racy situation.  The problem is that an endpoint can be closed (closing the rdma_cm_id) at the same time we try to read and process an event from the rdma_cm.  The current code protects this using eq->lock.  We hold the eq->lock when reading and processing rdma_cm events.  We also acquire the eq->lock when closing the ep and destroying the rdma_cm_id.

However, we hold the progress->ep_lock when reading and processing CQ entries.  If we see a failed transfer, we transition the EP into a disconnected state and write an entry to the EQ.  This requires a lock ordering of: progress->ep_lock -> eq->lock.  So, I can clearly not choose the wine in front of you.

But, the progress->ep_lock is needed to protect the state of the EP.  And the EP state can be driven by rdma_cm events.  This requires a lock ordering of: eq->lock -> progress->ep_lock.  So, I can clearly not choose the wine in front or me.

The current code has implemented the first lock ordering option: progress->ep_lock -> eq->lock,

shefty commented:
@zachdworkin - Can you please check the failures for this PR?  The failure is in trying to clone the repo, so maybe there's an issue with the proxy or something...

zachdworkin commented:
@shefty Fixed

shefty commented:
@swelch - These changes update the locking used by XRC.  XRC was using the eq->lock to protect some state information.  To avoid nested locking between the ep and eq, I added a new eq lock, an event_lock.  It's not a great name, but hopefully it's a temporary solution.  I added state checking to the ep, but did not try to integrate the xrc state into the new state.  It was easier not to touch it.

chien-intel commented on prov/verbs/src/verbs_ep.c:
```diff
@@ -114,26 +114,18 @@ ssize_t vrb_post_recv(struct vrb_ep *ep, struct ibv_recv_wr *wr)
-		return;
-	if (attr.cur_qp_state != IBV_QPS_ERR)
+	ofi_genlock_held(vrb_ep2_progress(ep)->active_lock);
+	if (!ep || ep->state != VRB_CONNECTED || !ep->eq)
```
ep is already used in the line above, no need to check it here.

shefty replied:
good catch... coverity would have complained about this.  I'll update

chien-intel replied:
when I get a chance I will run io500 with PR

chien-intel replied:
No performance regression with IO500/DAOS
